### DISCUSSION TITLE
Next-offline and WorkboxOpts: Error: "runtimeCaching" is not a supported parameter. How to enable "runtimeCaching" in next.config.js

### DISCUSSION THREAD

**Discussion by antonioOrtiz** [vercel/next.js#19906] [Help]
I'm just getting started with [next-offline][1] and found the section regarding workbox integration and its recipes. 

According to the docs:

> If you're new to workbox, I'd recommend reading this quick guide --
> anything inside of workboxOpts will be passed to
> workbox-webpack-plugin.
> 
> Define a workboxOpts object in your next.config.js and it will gets
> passed to workbox-webpack-plugin. Workbox is what next-offline uses
> under the hood to generate the service worker, you can learn more
> about it here.

After digging around, I found this great [section][2].

Essentially it gives a suggestion to use two different options:

[GenerateSW][3] or [InjectManifest][4]

I would like to use the InjectManifest, however when I try to implement that in my `next.config.js` file. I get this error:

    "runtimeCaching" is not a supported parameter.

This is my next.config.js:

    const withCSS = require('@zeit/next-css');
    const withSass = require('@zeit/next-sass');
    const withImages = require('next-images');
    const optimizedImages = require('next-optimized-images');
    const withOffline = require('next-offline');
    
    module.exports = withOffline(
      withImages(
        optimizedImages(
          withCSS(
            withSass({
              // useFileSystemPublicRoutes: false,
              // generateSw: false, // this allows all your workboxOpts to be passed in injectManifest
              generateInDevMode: true,
              workboxOpts: {
                swDest: './service-worker.js', // this is the important part,
                exclude: [/.+error\.js$/, /\.map$/, /\.(?:png|jpg|jpeg|svg)$/],
                runtimeCaching: [
                  {
                    urlPattern: /\.(?:png|jpg|jpeg|svg)$/,
                    handler: 'CacheFirst',
                    options: {
                      cacheName: 'hillfinder-images'
                    }
                  },
                  {
                    urlPattern: /^https?.*/,
                    handler: 'NetworkFirst',
                    options: {
                      cacheName: 'hillfinder-https-calls',
                      networkTimeoutSeconds: 15,
                      expiration: {
                        maxEntries: 150,
                        maxAgeSeconds: 30 * 24 * 60 * 60 // 1 month
                      },
                      cacheableResponse: {
                        statuses: [0, 200]
                      }
                    }
                  }
                ]
              },
              dontAutoRegisterSw: false,
              env: {
                MAPBOX_ACCESS_TOKEN: process.env.MAPBOX_ACCESS_TOKEN,
                useFileSystemPublicRoutes: false
              },
    
              webpack(config, options) {
                config.module.rules.push({
                  test: /\.(png|jpg|gif|svg|eot|ttf|woff|woff2)$/,
                  use: {
                    loader: 'url-loader',
                    options: {
                      limit: 100000,
                      target: 'serverless'
                    }
                  }
                });
    
                return config;
              }
            })
          )
        )
      )
    );

Also when I check the `Application` pane, in devTools I see this:

[![enter image description here][5]][5]

You'll notice what appears to me a duplication of fields i.e. `https-calls` and `hillfinder-https` and `images` and `hillfinder-images`. 

I thought the `cacheName` field in the `options: {}` in each was allowing one to include a custom name?

Just wondering if anyone has had experience setting this up?

Thank you in advance!

  [1]: https://github.com/hanford/next-offline
  [2]: https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin#which_plugin_to_use
  [3]: https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin#generatesw
  [4]: https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin#injectmanifest
  [5]: https://i.stack.imgur.com/MLYwS.jpg

**Comment by jerrygreen** [comment]
Not a direct answer to your question, but since you just started implementing this into your project, consider using `next-pwa` instead:

https://github.com/shadowwalker/next-pwa

I personally use it, with custom runtimeCaching (extending the default one, which is quite good, except that I need a different `start_url`, rather than `/`), and it works just fine.

They both use workbox under the hood, but to me next-pwa seems to be much a better option since it provides some great defaults, and it has several good examples of how to change different aspects of behavior (like custom worker, etc.), similar to the approach used by NextJS
